scheduler/test/cook/test/simulator.clj

```diff
@@ -0,0 +1,571 @@
+        state-atom (atom {})
+        make-mesos-driver-fn (fn [scheduler _];; _ is framework-id
+                               (mm/mesos-mock mesos-hosts offer-trigger-chan scheduler
+                                              :task->runtime-ms task->runtime-ms
+                                              :task->complete-status task->complete-status
+                                              :complete-trigger-chan complete-trigger-chan
+                                              :state-atom state-atom))
+        initial-time (System/currentTimeMillis)
+        config (merge {:shares [{:user "default" :mem 4000.0 :cpus 4.0 :gpus 1.0}]}
+                      config)
+        scheduler-config (merge (:scheduler-config config)
+                                {:trigger-chans {:rank-trigger-chan ranker-trigger-chan
+                                                 :match-trigger-chan matcher-trigger-chan
+                                                 :rebalancer-trigger-chan rebalancer-trigger-chan
+                                                 ;; Don't care about these yet
+                                                 :straggler-trigger-chan (async/chan)
+                                                 :lingering-task-trigger-chan (async/chan)
+                                                 :cancelled-task-trigger-chan (async/chan)}})]
+    ;; We are setting time to enable us to have deterministic runs
+    ;; of the simulator while hooking into the scheduler as non-invasively
+    ;; as possible. A longer explanation can be found in the simulator dev docs.
+    (org.joda.time.DateTimeUtils/setCurrentMillisFixed initial-time)
+    (log/info "Starting simulation.")
+    (with-cook-scheduler mesos-datomic-conn make-mesos-driver-fn
+      scheduler-config
+      (try
+        (doseq [{:keys [user mem cpus gpus]} (:shares config)]
+          (share/set-share! mesos-datomic-conn user "simulation" :mem mem :cpus cpus :gpus gpus))
+        (loop [trace trace
+               simulation-time simulation-time
+               fake-real-time initial-time]
+
+          (org.joda.time.DateTimeUtils/setCurrentMillisFixed fake-real-time)
+
+          (let [start-ms (System/currentTimeMillis)
+                submission-batch (take-while #(<= (:submit-time-ms %) simulation-time) trace)
+                send-offers-complete-chan (async/chan)
+                flush-complete-chan (async/chan)
+                match-complete-chan (async/chan)
+                rank-complete-chan (async/chan)
+                rebalancer-complete-chan (async/chan)]
+            (log/info "Simulation time: " simulation-time)
+            (log/info "submission-batch size: " (count submission-batch))
+
+            (log/info "Submitting batch")
+            ;; Submit new jobs
+            (doseq [job submission-batch]
+              (org.joda.time.DateTimeUtils/setCurrentMillisFixed (inc (.getTime (tc/to-date (t/now)))))
```

Comment:
Why does this call have to be inside the loop? Is it a problem if all the jobs have the same submission time?